src/test/script_segwit_tests.cpp

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2012-2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <script/script.h>
+#include <test/test_bitcoin.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_FIXTURE_TEST_SUITE(script_segwit_tests, BasicTestingSetup)
+
+BOOST_AUTO_TEST_CASE(IsPayToWitnessScriptHash_Valid)
+{
+    uint256 dummy;
+    CScript p2wsh;
+    p2wsh << OP_0 << ToByteVector(dummy);
+    BOOST_CHECK(p2wsh.IsPayToWitnessScriptHash());
+
+    std::vector<unsigned char> bytes = {OP_0, 32};
+    bytes.insert(bytes.end(), 32, 0);
+    BOOST_CHECK(CScript(bytes.begin(), bytes.end()).IsPayToWitnessScriptHash());
+}
+
+BOOST_AUTO_TEST_CASE(IsPayToWitnessScriptHash_Invalid_NotOp0)
+{
+    uint256 dummy;
+    CScript notp2wsh;
+    notp2wsh << OP_1 << ToByteVector(dummy);
+    BOOST_CHECK(!notp2wsh.IsPayToWitnessScriptHash());
+}
+
+BOOST_AUTO_TEST_CASE(IsPayToWitnessScriptHash_Invalid_Size)
+{
+    uint160 dummy;
+    CScript notp2wsh;
+    notp2wsh << OP_0 << ToByteVector(dummy);
+    BOOST_CHECK(!notp2wsh.IsPayToWitnessScriptHash());
+}
+
+BOOST_AUTO_TEST_CASE(IsPayToWitnessScriptHash_Invalid_Nop)
+{
+    uint256 dummy;
+    CScript notp2wsh;
+    notp2wsh << OP_0 << OP_NOP << ToByteVector(dummy);
+    BOOST_CHECK(!notp2wsh.IsPayToWitnessScriptHash());
+}
+
+BOOST_AUTO_TEST_CASE(IsPayToWitnessScriptHash_Invalid_EmptyScript)
+{
+    CScript notp2wsh;
+    BOOST_CHECK(!notp2wsh.IsPayToWitnessScriptHash());
+}
+
+BOOST_AUTO_TEST_CASE(IsPayToWitnessScriptHash_Invalid_Pushdata)
+{
+    // A script is not P2WSH if OP_PUSHDATA is used to push the hash.
+    std::vector<unsigned char> bytes = {OP_0, OP_PUSHDATA1, 32};
+    bytes.insert(bytes.end(), 32, 0);
+    BOOST_CHECK(!CScript(bytes.begin(), bytes.end()).IsPayToWitnessScriptHash());
+
+    bytes = {OP_0, OP_PUSHDATA2, 32, 0};
+    bytes.insert(bytes.end(), 32, 0);
+    BOOST_CHECK(!CScript(bytes.begin(), bytes.end()).IsPayToWitnessScriptHash());
+
+    bytes = {OP_0, OP_PUSHDATA4, 32, 0, 0, 0};
+    bytes.insert(bytes.end(), 32, 0);
+    BOOST_CHECK(!CScript(bytes.begin(), bytes.end()).IsPayToWitnessScriptHash());
+}
+
+namespace {
+
+bool IsExpectedWitnessProgram(const CScript& script, const int expectedVersion, const std::vector<unsigned char>& expectedProgram)
+{
+    int actualVersion;
+    std::vector<unsigned char> actualProgram;
+    if (!script.IsWitnessProgram(actualVersion, actualProgram)) {
+        return false;
+    }
+    BOOST_CHECK_EQUAL(actualVersion, expectedVersion);
+    BOOST_CHECK(actualProgram == expectedProgram);
+    return true;
+}
+
+bool IsNoWitnessProgram(const CScript& script)
+{
+    int dummyVersion;
+    std::vector<unsigned char> dummyProgram;
+    return !script.IsWitnessProgram(dummyVersion, dummyProgram);
+}
+
+} // anonymous namespace
+
+BOOST_AUTO_TEST_CASE(IsWitnessProgram_Valid)
+{
+    std::vector<unsigned char> program = {42, 18};
```

Comment:
```suggestion
    // Witness programs have a minimum data push of 2 bytes
    std::vector<unsigned char> program = {42, 18};
```

Reply:
Added to the commit, thanks for the suggestion!